### ISSUE TITLE
Queries never complete

### ISSUE THREAD

**Issue by steven-sheehy** [r2dbc/r2dbc-postgresql#247]
## Bug Report

Reproduction project: https://github.com/steven-sheehy/r2dbc-stalled-clients

If connection pool size is set to X, then whenever more than X subscribers run a query concurrently they are stalled and never receive any messages.

#### Versions

- Driver: 0.8.2.BUILD-SNAPSHOT
- Database: Postgres
- Java: 11
- OS: MacOS

#### Current Behavior

1. Set maxPoolSize to 10
2. Start 15 concurrent queries
3. 5 queries never finish even after the other 10 finish and never receive an error

<details>
<summary>Stack trace</summary>

```
// your stack trace here
```

</details>
 
#### Table schema

<details>
<summary>Input Code</summary>

```sql
-- your SQL here;
```

</details>

#### Steps to reproduce

<details>
<summary>Input Code</summary>

```java
// your code here;
```

</details>

#### Expected behavior/code

1. Set maxPoolSize to 10
2. Start 15 concurrent queries
3. All 15 queries finish eventually with 5 being queued for a short period until a connection becomes available. 

#### Possible Solution

- Queue queries when connection pool exhausted up to a reasonable amount of number/time
- Throw a connection pool exhausted if the number of queries exceeds queue size or any single query is in queue more than some timeout

**Comment by steven-sheehy** [comment]
Also occurs in 0.8.0

**Comment by mp911de** [comment]
What do you mean with more queries/subscribers than pool size?

The pool hands out only as many connections as there are configured. Any subsequent allocation is enqueued and awaits a free connection. A bit of sample code would be helpful.

**Comment by mp911de** [comment]
I'm no longer able to reproduce the issue with 0.8.2.BUILD-SNAPSHOT. I can reproduce the issue with 0.8.0.RELEASE and 0.8.1.RELEASE.

Can you give it a retry to check whether we can close this ticket?

**Comment by steven-sheehy** [comment]
Yes, with `0.8.2.BUILD-20200214.102436-4` it now works. Thanks for checking.

**Comment by steven-sheehy** [comment]
I'm reopening this as I experienced the issue with `0.8.2.BUILD-20200228.083734-6`. I've updated the test repo to fail with this version. The test case is a bit convoluted as I don't know how to test multiple subscribers concurrently very well.

The test case is 3 subscribers each polling the database repeatedly to get the next 1000 messages at a time. Additionally, I simulate client backpressure using sleep in on next and use publishOn and subscribeOn. Without the sleep, publishOn or subscribeOn the issue does not seem to occur. This scenario is my attempt to reproduce my company's scenario with paging this relational data via a streaming GRPC API.

I no longer think it's related to exceeding max connections but is instead some race condition with multiple subscribers when back pressure occurs. It seems like at this time it switches from publishing on reactor epoll threads to the subscription thread (GRPC executor pool in my case) and that's around when the issue appears.

**Comment by mp911de** [comment]
Thanks Steven. This issue requires further investigation of what happens and why it happens. I also tend towards a cause that isn't caught properly with either subscription (demand) or concurrency handling.

**Comment by steven-sheehy** [comment]
Not sure why this is classified as a task and ideal for contribution. In my opinion, it's a P1 bug that is hard for newbies like myself to track down. R2DBC never recovers from this scenario without a JVM restart. It makes R2DBC unusable for our use case and I was forced to look into to migrating away from the library.

**Comment by mp911de** [comment]
The reason this ticket is labeled with `ideal for contribution` is rooted in the fact that all maintainers on this project are busy with some other kind of work (mostly their day job) and we ask the community for additional help. 

I'm sorry to hear about the hassle you're having with that issue. This specific issue requires a couple of days of investigation and a clear mind which wasn't possible at least for me during the past days given the dynamic situation the world faces currently.

**Comment by Squiry** [comment]
I've [reproduced ](https://gist.github.com/Squiry/3b88bffa0a502e771555d8317a589cc7) this issue with pure r2dbc-postgres.
The problem is in `.publishOn(Schedulers.boundedElastic())` that line.
Usually we have our messages on reactor-tcp-nio threads, but once in a while it is published on bounded elastic scheduller (I don't know why). In that case `PostgresqlResult`' sends `onComplete`, but underlying `Flux<PostgresqlResult>.flatMap` doesn't receive that signal. The logic is somewhere near `FluxFlatMap.FlatMapMain.innerComplete`: the complete signal should have been sent from `drainLoop` which wasn't called. And I'm not that familiar with reactor-core sources to tell what's the problem.

**Comment by mp911de** [comment]
Can you push the code somewhere? I will ask @bsideup or @simonbasle for help.

**Comment by bsideup** [comment]
@Squiry the reproducer uses uninterruptible `sleep`:
https://gist.github.com/Squiry/3b88bffa0a502e771555d8317a589cc7#file-r2dbcexampleapplicationtests-L182

it does not "simulate backpressure" but blocks the IO thread.
Consider using `delayElements` instead, or prepending it with `publishOn`.

**Comment by mp911de** [comment]
I wasn't able to reproduce the issue. In general, the provided reproducer issues uses a number of rows that make it almost impossible to debug the code. I understand that race conditions are somewhat tricky to reproduce so my hope is that we can find another way to reproduce the issue with less test complexity involved.

**Comment by Squiry** [comment]
Computation is blocking operation too, you know. Problem is not in blocking, 

Anyway, I have my ide paused in debug session with a request that going to hang and I'd appreciate some help. Stack traces are 1km long but: 

-  [here](https://i.imgur.com/bR99ihR.png)'s usual process of frame: it goes through all the reactor subscribers to that onNext with sleep.
-  [here](https://i.imgur.com/X8FDIhX.png)'s process of frame, after group of wich we'll get onComplete. Notice the thread. That's happening once in a 30k frames.
-  And [here](https://i.imgur.com/oTquDep.png)'s the CommandComplete frame, that will lead to onComplete() in flux.handle, but client's subscriber will never get it.
I'm going to spend some time on weekend, but I think that I'm missing something. Still don't get how the stack of subscribers is different for the same code: first one has real subscirber (onNext) after group of FluxPeek, and the second has timeout and publishOn subscribers.

**Comment by mp911de** [comment]
@steven-sheehy we fixed an issue in the context of #285 whose symptoms were that queries never complete. Care to test against 0.8.4.BUILD-SNAPSHOT whether the issue is gone?

**Comment by steven-sheehy** [comment]
My real application no longer uses R2DBC, so I can't test it there. I did try again with my sample project and it looks to be fixed. Will close. Thanks!
